tests/fault_tolerance/etcd_ha/utils.py

```
@@ -156,123 +161,217 @@ def __init__(
         for i, replica in enumerate(self.replicas):
             if replica:  # Only include active replicas
-                client_port = self.base_client_port + i
-                endpoints.append(f"http://localhost:{client_port}")
+                client_port = self.base_port + (2 * i)
+                endpoints.append(f"http://127.0.0.1:{client_port}")
         return endpoints
 
+    def terminate_replica(self, idx: int):
+        """Terminate a specific replica by index."""
+        if idx < 0 or idx >= self.num_replicas:
+            raise RuntimeError(f"Invalid replica index: {idx}")
```

Ruff (0.14.4)

Avoid specifying long messages outside the exception class

(TRY003)